src/hotspot/cpu/x86/universalUpcallHandler_x86_64.cpp

```diff
@@ -281,38 +168,16 @@ static void restore_callee_saved_registers(MacroAssembler* _masm, const ABIDescr
 // "0" is assigned for rax and for xmm0. Thus we need to ignore -Wnonnull.
 PRAGMA_DIAG_PUSH
 PRAGMA_NONNULL_IGNORED
-address ProgrammableUpcallHandler::generate_optimized_upcall_stub(jobject receiver, Method* entry, jobject jabi, jobject jconv) {
-  ResourceMark rm;
+address ProgrammableUpcallHandler::generate_optimized_upcall_stub(jobject receiver, Method* entry,
+                                                                  BasicType* in_sig_bt, int total_in_args,
+                                                                  BasicType* out_sig_bt, int total_out_args,
+                                                                  BasicType ret_type,
+                                                                  jobject jabi, jobject jconv,
+                                                                  bool needs_return_buffer, int ret_buf_size) {
   const ABIDescriptor abi = ForeignGlobals::parse_abi_descriptor(jabi);
   const CallRegs call_regs = ForeignGlobals::parse_call_regs(jconv);
-  assert(call_regs._rets_length <= 1, "no multi reg returns");
   CodeBuffer buffer("upcall_stub_linkToNative", /* code_size = */ 2048, /* locs_size = */ 1024);
 
-  int register_size = sizeof(uintptr_t);
-  int buffer_alignment = xmm_reg_size;
-
-  assert(entry->is_static(), "static only");
-  // Fill in the signature array, for the calling-convention call.
-  const int total_out_args = entry->size_of_parameters();
-  assert(total_out_args > 0, "receiver arg");
-
-  BasicType* out_sig_bt = NEW_RESOURCE_ARRAY(BasicType, total_out_args);
-  BasicType ret_type;
-  {
-    int i = 0;
-    SignatureStream ss(entry->signature());
-    for (; !ss.at_return_type(); ss.next()) {
-      out_sig_bt[i++] = ss.type();  // Collect remaining bits of signature
-      if (ss.type() == T_LONG || ss.type() == T_DOUBLE)
-        out_sig_bt[i++] = T_VOID;   // Longs & doubles take 2 Java slots
-    }
-    assert(i == total_out_args, "");
-    ret_type = ss.type();
-  }
-  // skip receiver
-  BasicType* in_sig_bt = out_sig_bt + 1;
-  int total_in_args = total_out_args - 1;
```

Comment:
I've moved this to the shared code in universalUpcallHanlder.cpp, since it can be shared between platforms.